### PR TITLE
Add OTNS unit tests

### PR DESCRIPTION
**Pull request by ctan-g**
This PR adds OTNS unit tests to Silk.

- Add unit test related mock services.
- Add OTNS manager unit tests.
- Add silk log replayer unit tests.
  - Add log replaying related fixture files.
- Minor formatting fixes.
- Fix one issue found in writing tests: `testcase` calling `set_test_title` after a test is run.

### CODE REVIEW

File: silk/tests/silk_replay.py [hunk 1]
```diff
@@ -264,12 +272,22 @@ def output_summary(self, coalesced: bool, csv_path: str):
...
-    def run(self):
+    def run(self, start_line: int = 0, stop_regex: str = None):
         """Run the Silk log replayer.
+        
+        This method provides two optional arguments to allow for unit testing.
+        
+        Args:
+            start_line (int, optional): start reading the log file at the specified line number. Defaults to 0.
+            stop_regex (str, optional): stop running if the pattern matches a log line. Defaults to None.
         """
```
Comment (qilinkang) [inline silk/tests/silk_replay.py:286]: Pls add Returns:

File: silk/unit_tests/test_log_replay.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        self.exception_queue = queue.Queue()
+
+        self.manager = OtnsManager("localhost", self.logger.getChild("OtnsManager"))
+        self.grpc_client = MockGrpcClient(self.exception_queue, self.logger.getChild("MockGrpcClient"))
+        self.manager.grpc_client = self.grpc_client
+
+        self.udp_server = MockUDPServer(self.exception_queue)
+
+        hwconfig_path = Path(__file__).parent / "fixture/hwconfig.ini"
```
Comment (qilinkang) [inline silk/unit_tests/test_log_replay.py:37]: pls check this line

File: silk/unit_tests/test_log_replay.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    def create_replayer(self, log_filename: str) -> silk_replay.SilkReplayer:
+        """Prepare a replayer for a test.
+
+        Args:
+            log_filename (str): log file basename.
+
+        Returns:
+            SilkReplayer: a SilkReplayer prepared for a test.
+        """
+        log_path = str(Path(__file__).parent / f"fixture/{log_filename}")
```
Comment (qilinkang) [inline silk/unit_tests/test_log_replay.py:49]: Pls check this line

File: silk/unit_tests/test_log_replay.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return replayer
+
+    def testReplayFormNetwork(self):
+        """Test replaying the form network test case log.
+        """
+        replayer = self.create_replayer("form_network_log.txt")
+        # setting up
+        line_number = replayer.run(stop_regex=r"Sent cmd: title \"TestFormNetwork.set_up\"")
+        # pairing
+        line_number = replayer.run(start_line=line_number,
```
Comment (qilinkang) [inline silk/unit_tests/test_log_replay.py]: How do we decide if the replay is a Pass or Fail ?

File: silk/unit_tests/test_log_replay.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        self.exception_queue = queue.Queue()
+
+        self.manager = OtnsManager("localhost", self.logger.getChild("OtnsManager"))
+        self.grpc_client = MockGrpcClient(self.exception_queue, self.logger.getChild("MockGrpcClient"))
+        self.manager.grpc_client = self.grpc_client
+
+        self.udp_server = MockUDPServer(self.exception_queue)
+
+        hwconfig_path = Path(__file__).parent / "fixture/hwconfig.ini"
+        self.args = ["tests/silk_replay.py", "-v2", "-c", str(hwconfig_path), "-s", "localhost", "-p", "100"]
```
Comment (qilinkang) [inline silk/unit_tests/test_log_replay.py:38]: Is speed of 100 too much ?

File: silk/unit_tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+def random_string(length: int) -> str:
+    """Generate a random string with defined length.
+
+    Args:
+        length (int): desired string length.
+
+    Returns:
+        str: generated random string.
+    """
+    return ''.join(random.choice(string.ascii_letters) for i in range(length))
```
Comment (qilinkang) [inline silk/unit_tests/test_utils.py]: for _ in range(length)

File: silk/unit_tests/test_log_replay.py (hunk 4 above)
Comment (ctan-g) [reply silk/unit_tests/test_log_replay.py]: I am adding assertion to the replay tests. The replay tests are WIP.

File: silk/unit_tests/test_log_replay.py (hunk 2 above)
Comment (ctan-g) [reply silk/unit_tests/test_log_replay.py:37]: This is a correct syntax for the `pathlib` usage. See https://docs.python.org/3/library/pathlib.html

File: silk/unit_tests/test_log_replay.py (hunk 3 above)
Comment (ctan-g) [reply silk/unit_tests/test_log_replay.py:49]: This is also a correct syntax for the `pathlib` usage. See https://docs.python.org/3/library/pathlib.html

File: silk/unit_tests/test_log_replay.py (hunk 5 above)
Comment (ctan-g) [reply silk/unit_tests/test_log_replay.py:38]: If the speed is lower the unit tests are going to take longer. I think 100 is an appropriate setting.

File: silk/unit_tests/test_log_replay.py [hunk 7]
```diff
@@ -50,7 +50,7 @@ def testReplayFormNetwork(self):
         log_path = self.get_log_path("form_network_log.txt")
         with log_path.open() as log:
             for line in log:
-                
+                pass
```
Comment (qilinkang) [inline silk/unit_tests/test_log_replay.py]: Seems missing the test ?

**Review by qilinkang — COMMENTED** [review]
Please post the test run result or log.

File: silk/unit_tests/test_log_replay.py (hunk 7 above)
Comment (ctan-g) [reply silk/unit_tests/test_log_replay.py]: This was a placeholder in the older commits.

**Comment by ctan-g** [comment]
> Please post the test run result or log.

Should be able to see them in the detail page of the CI checks, for example here: https://github.com/openthread/silk/runs/1040005080?check_suite_focus=true

**Review by qilinkang — APPROVED** [review]
LGTM

**Review by jwhui — APPROVED** [review]

**Pull request merged**